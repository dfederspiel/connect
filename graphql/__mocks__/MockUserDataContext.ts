import { IUserDataContext } from '@lib/auth/UserDataContext';
import { User } from '@prisma/client';

const user = {
  id: 1,
  domain: 'codefly.ninja',
  email: 'david@codefly.ninja',
} as User;

export const MockUserDataContext = {
  createUser: async () => user,
  getById: async () => user,
  getAll: async () => [user],
  getByEmail: async () => user,
  post: async (user: User) => user,
} as IUserDataContext;
