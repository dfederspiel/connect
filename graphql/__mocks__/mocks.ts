import { User } from '@prisma/client';
import { IMocks, MockList } from '@graphql-tools/mock';
import { faker } from '@faker-js/faker';
export interface IConnectMocks extends IMocks {
  User(): User;
}

export const DataMocks = {
  Int: () => faker.datatype.number(100),
  Float: () => faker.datatype.float(10),
  String: () => faker.random.words(5),
  User: () => ({
    id: faker.datatype.number(100),
    domain: faker.internet.domainName(),
    email: faker.internet.email(),
    skills: () => new MockList([0, 12]),
  }),
  Query: () => ({
    users: () => new MockList([0, 12]),
  }),
} as IConnectMocks;
