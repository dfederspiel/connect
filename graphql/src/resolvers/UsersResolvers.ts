import { User } from '@prisma/client';
import { gql, IResolvers } from 'apollo-server-express';
import UserDataSource from '../datasources/UsersDataSource';

export const UsersTypeDefs = gql`
  type User {
    id: ID!
    email: String
    domain: String
  }

  extend type Query {
    user: User
    users: [User]
  }
`;

export default class UsersResolvers {
  resolvers: IResolvers;

  constructor() {
    this.resolvers = this.initializeResolvers();
  }

  initializeResolvers = (): IResolvers => {
    return {
      Query: {
        user: (
          _: any,
          _args: any,
          context: { user: User; dataSources: { userApi: UserDataSource } },
          __: any,
        ) => {
          if (!context.user)
            return {
              error: 'There is no user context, did you forget to pass a bearer token?',
            };
          return context.dataSources.userApi.getById(context.user.id.toString());
        },
        users: (
          _: any,
          _args: any,
          context: { user: User; dataSources: { userApi: UserDataSource } },
          __: any,
        ) => {
          if (!context.user)
            return {
              error: 'There is no user context, did you forget to pass a bearer token?',
            };
          return context.dataSources.userApi.getAll();
        },
      },
    };
  };
}
