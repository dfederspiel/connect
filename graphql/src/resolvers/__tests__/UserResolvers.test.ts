import { User } from '@prisma/client';
import { MockUserDataContext } from '../../../__mocks__/MockUserDataContext';
import UsersResolvers from '../UsersResolvers';

const user = {
  id: 1,
  domain: '',
  email: 'david@codefly.ninja',
};

const context = {
  dataSources: {
    users: {
      getById: async (): Promise<User | null> => user,
      getAll: async (): Promise<User[]> => [user, user],
      context: MockUserDataContext,
    } as any,
  },
  user: {
    id: 1,
    domain: '',
    email: '',
  } as User,
};

describe('the user resolvers', () => {
  it('exists', () => {
    const resolver = new UsersResolvers();
    expect(resolver).toBeDefined();
  });

  describe('when the user context is set', () => {
    it('can get the current user', async () => {
      const resolver = new UsersResolvers();
      const user = await resolver.resolvers.Query.user(null, null, context, null);
      expect(user).toMatchSnapshot();
    });

    it('can resolve all users', async () => {
      const resolver = new UsersResolvers();
      const users = await resolver.resolvers.Query.users(null, null, context, null);
      expect(users).toMatchSnapshot();
    });
  });
});
