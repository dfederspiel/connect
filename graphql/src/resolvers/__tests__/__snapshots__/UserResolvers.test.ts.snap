// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the user resolvers when the user context is set can get the current user 1`] = `
Object {
  "domain": "",
  "email": "david@codefly.ninja",
  "id": 1,
}
`;

exports[`the user resolvers when the user context is set can resolve all users 1`] = `
Array [
  Object {
    "domain": "",
    "email": "david@codefly.ninja",
    "id": 1,
  },
  Object {
    "domain": "",
    "email": "david@codefly.ninja",
    "id": 1,
  },
]
`;
